lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp

```diff
@@ -195,6 +195,87 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
                       binder.op, resultType, operand);
                   return success();
                 });
+  patterns.onOp(
+      "LogSoftmax", 13,
+      [](OpBinder binder, ConversionPatternRewriter &rewriter) {
+        Value input;
+        Torch::ValueTensorType resultType;
+        if (binder.tensorOperand(input) || binder.tensorResultType(resultType))
+          return failure();
+        int64_t axis;
+        if (binder.s64IntegerAttr(axis, "axis", -1))
+          return rewriter.notifyMatchFailure(binder.op, "axis bind failure");
+        Value axisConst = rewriter.create<Torch::ConstantIntOp>(
+            binder.getLoc(), rewriter.getI64IntegerAttr(axis));
+        Value none = rewriter.create<Torch::ConstantNoneOp>(binder.getLoc());
+        rewriter.replaceOpWithNewOp<Torch::AtenLogSoftmaxIntOp>(
+            binder.op, resultType, input, axisConst, none);
+        return success();
+      });
+  patterns.onOp(
+      "LogSoftmax", 1,
+      [](OpBinder binder, ConversionPatternRewriter &rewriter) {
+        Value input;
+        Torch::ValueTensorType resultType;
+        if (binder.tensorOperand(input) || binder.tensorResultType(resultType))
+          return failure();
+
+        int64_t axis;
+        if (binder.s64IntegerAttr(axis, "axis", 1))
+          return rewriter.notifyMatchFailure(binder.op, "axis bind failure");
+        if (axis < 0) {
+          return rewriter.notifyMatchFailure(
+              binder.op, "Not yet implemented: negative axis.");
+        }
+        std::optional<unsigned> maybeRank = Torch::getTensorRank(input);
+        if (!maybeRank)
+          return rewriter.notifyMatchFailure(binder.op,
+                                             "Unsupported: unranked tensor");
+        int64_t rank = *maybeRank;
+        auto inputTy = input.getType().cast<Torch::ValueTensorType>();
+        if (!inputTy || !inputTy.hasSizes())
+          return rewriter.notifyMatchFailure(
+              binder.op, "failed to get input type or sizes");
+
+        Value axisConst = rewriter.create<Torch::ConstantIntOp>(
+            binder.getLoc(), rewriter.getI64IntegerAttr(axis));
+        Value none = rewriter.create<Torch::ConstantNoneOp>(binder.getLoc());
+        Value cstEnd = rewriter.create<Torch::ConstantIntOp>(
+            binder.getLoc(), rewriter.getI64IntegerAttr(rank - axis));
+
+        if (!inputTy || !inputTy.hasSizes())
+          return failure();
+        llvm::SmallVector<int64_t> allDims(inputTy.getSizes());
+        llvm::SmallVector<int64_t> leftDims(allDims.begin(),
+                                            allDims.begin() + axis);
+        llvm::SmallVector<int64_t> rightDims(allDims.begin() + axis,
```

Comment:
You can make a `llvm::ArrayRef` for these. It works as a pair of pointers and lets you track it.

Reply:
Oh, let me try that out right now.

Reply:
I kept leftDims as a vector (because I need to append the product of right dims to the end) but converted the others to ArrayRefs and reorganized for readability.